File: trunk/jung/jung-visualization/src/main/java/edu/uci/ics/jung/visualization/control/SatelliteShearingGraphMousePlugin.java
/*
 * Copyright (c) 2005, the JUNG Project and the Regents of the University of
 * California All rights reserved.
 *
 * This software is open-source under the BSD license; see either "license.txt"
 * or http://jung.sourceforge.net/license.txt for a description.
 *
 * Created on Aug 15, 2005
 */

package edu.uci.ics.jung.visualization.control;

import java.awt.Dimension;
import java.awt.event.MouseEvent;
import java.awt.geom.Point2D;

import edu.uci.ics.jung.visualization.Layer;
import edu.uci.ics.jung.visualization.VisualizationViewer;
import edu.uci.ics.jung.visualization.transform.MutableTransformer;

/**
 * Overrides ShearingGraphMousePlugin so that mouse events in the
 * satellite view cause shearing of the main view
 * 
 * @see ShearingGraphMousePlugin
 * @author Tom Nelson 
 *
 */
public class SatelliteShearingGraphMousePlugin extends ShearingGraphMousePlugin {

    public SatelliteShearingGraphMousePlugin() {
        super();
    }

    public SatelliteShearingGraphMousePlugin(int modifiers) {
        super(modifiers);
    }
    
    /**
     * overridden to shear the main view
     */
    public void mouseDragged(MouseEvent e) {
        if(down == null) return;
        VisualizationViewer vv = (VisualizationViewer)e.getSource();
        boolean accepted = checkModifiers(e);
        if(accepted) {
            if(vv instanceof SatelliteVisualizationViewer) {
                VisualizationViewer vvMaster = 
                    ((SatelliteVisualizationViewer)vv).getMaster();
                
                MutableTransformer modelTransformerMaster = 
                	vvMaster.getRenderContext().getMultiLayerTransformer().getTransformer(Layer.LAYOUT);

                vv.setCursor(cursor);
                Point2D q = down;
                Point2D p = e.getPoint();
                float dx = (float) (p.getX()-q.getX());
                float dy = (float) (p.getY()-q.getY());

                Dimension d = vv.getSize();
                float shx = 2.f*dx/d.height;
                float shy = 2.f*dy/d.width;
                // I want to compute shear based on the view coordinates of the
                // lens center in the satellite view.
                // translate the master view center to layout coords, then translate
                // that point to the satellite view's view coordinate system....
                Point2D center = vv.getRenderContext().getMultiLayerTransformer().transform(vvMaster.getRenderContext().getMultiLayerTransformer().inverseTransform(vvMaster.getCenter()));
                if(p.getX() < center.getX()) {
                    shy = -shy;
                }
                if(p.getY() < center.getY()) {
                    shx = -shx;
                }
                modelTransformerMaster.shear(-shx, -shy, vvMaster.getCenter());

                down.x = e.getX();
                down.y = e.getY();
            }
            e.consume();
        }
    }
}

File: jung-visualization/src/main/kotlin/edu/uci/ics/jung/visualization/control/SatelliteShearingGraphMousePlugin.kt
/*
 * Copyright (c) 2005, The JUNG Authors
 * All rights reserved.
 *
 * This software is open-source under the BSD license; see either "license.txt"
 * or https://github.com/jrtom/jung/blob/master/LICENSE for a description.
 *
 * Created on Aug 15, 2005
 */

package edu.uci.ics.jung.visualization.control

import edu.uci.ics.jung.visualization.MultiLayerTransformer.Layer
import edu.uci.ics.jung.visualization.VisualizationViewer
import java.awt.event.MouseEvent

/**
 * Overrides ShearingGraphMousePlugin so that mouse events in the satellite view cause shearing of
 * the main view
 *
 * @see ShearingGraphMousePlugin
 * @author Tom Nelson
 */
open class SatelliteShearingGraphMousePlugin : ShearingGraphMousePlugin {

    constructor() : super()

    constructor(modifiers: Int) : super(modifiers)

    /** overridden to shear the main view */
    override fun mouseDragged(e: MouseEvent) {
        if (down == null) {
            return
        }
        val vv = e.source as VisualizationViewer<*, *>
        val accepted = checkModifiers(e)
        if (accepted) {
            if (vv is SatelliteVisualizationViewer<*, *>) {
                val vvMaster = vv.master

                val modelTransformerMaster =
                    vvMaster.getRenderContext().getMultiLayerTransformer().getTransformer(Layer.LAYOUT)

                vv.cursor = cursor
                val q = down!!
                val p = e.point
                val dx = (p.x - q.x).toFloat()
                val dy = (p.y - q.y).toFloat()

                val d = vv.size
                var shx = 2f * dx / d.height
                var shy = 2f * dy / d.width
                // I want to compute shear based on the view coordinates of the
                // lens center in the satellite view.
                // translate the master view center to layout coords, then translate
                // that point to the satellite view's view coordinate system....
                val center = vv.getRenderContext().getMultiLayerTransformer().transform(
                    vvMaster.getRenderContext().getMultiLayerTransformer()
                        .inverseTransform(vvMaster.getCenter())
                )!!
                if (p.x < center.x) {
                    shy = -shy
                }
                if (p.y < center.y) {
                    shx = -shx
                }
                modelTransformerMaster.shear((-shx).toDouble(), (-shy).toDouble(), vvMaster.getCenter())

                down!!.x = e.x
                down!!.y = e.y
            }
            e.consume()
        }
    }
}
